crates/polars-parquet/src/parquet/read/page/memreader.rs

```diff
@@ -0,0 +1,191 @@
+use std::io;
+use std::ops::Deref;
+use std::sync::Arc;
+
+/// A cursor over a segment of heap allocated memory. This is used for the Parquet reader to avoid
+/// sequential allocations.
+#[derive(Debug, Clone)]
+pub struct MemReader {
+    data: Arc<[u8]>,
+    position: usize,
+}
+
+/// A reference to a slice of a memory reader.
+///
+/// This should not outlast the original the original [`MemReader`] because it still owns all the
+/// memory.
+#[derive(Debug, Clone)]
+pub struct MemReaderSlice {
+    data: Arc<[u8]>,
+    start: usize,
+    end: usize,
+}
+
+impl Default for MemReaderSlice {
+    fn default() -> Self {
+        let slice: &[u8] = &[];
+        Self {
+            data: Arc::from(slice),
+            start: 0,
+            end: 0,
+        }
+    }
+}
+
+impl Deref for MemReaderSlice {
+    type Target = [u8];
+
+    #[inline(always)]
+    fn deref(&self) -> &Self::Target {
+        &self.data[self.start..self.end]
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum CowBuffer {
+    Borrowed(MemReaderSlice),
+    Owned(Vec<u8>),
+}
+
+impl Deref for CowBuffer {
+    type Target = [u8];
+
+    #[inline(always)]
+    fn deref(&self) -> &Self::Target {
+        match self {
+            CowBuffer::Borrowed(v) => v.deref(),
+            CowBuffer::Owned(v) => v.deref(),
+        }
+    }
+}
+
+impl MemReader {
+    #[inline(always)]
+    pub fn new(data: Arc<[u8]>) -> Self {
+        Self { data, position: 0 }
+    }
+
+    #[inline(always)]
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    #[inline(always)]
+    pub fn remaining_len(&self) -> usize {
+        self.data.len() - self.position
+    }
+
+    #[inline(always)]
+    pub fn position(&self) -> usize {
+        self.position
+    }
+
+    #[inline(always)]
+    pub fn from_slice(data: &[u8]) -> Self {
+        let data = data.into();
+        Self { data, position: 0 }
+    }
+
+    #[inline(always)]
+    pub fn from_vec(data: Vec<u8>) -> Self {
+        let data = data.into_boxed_slice().into();
+        Self { data, position: 0 }
+    }
+
+    #[inline(always)]
+    pub fn from_reader<R: io::Read>(mut reader: R) -> io::Result<Self> {
+        let mut vec = Vec::new();
+        reader.read_to_end(&mut vec)?;
```

Comment:
FWIW can we mmap the file?

Maybe we can reuse the logic in `get_reader_bytes`.

Reply:
The idea is that we don't trigger IO reads anymore in our parquet reader. This is needed for the new streaming engine. We will read the needed row-groups/memory chunks with tokio and then offload to the parquet decoder which is compute bound.

Reply:
We can set [`populate()`](https://docs.rs/memmap2/latest/memmap2/struct.MmapOptions.html#method.populate) to avoid page fault.

Reply:
Yeah.. I do agree we should memory map for the default engine.

But I am not sure this is the location. This PR is to remove the `Read` trait out of parquet reader and then we can decide how to go into the reader on a per engine base, be it memory map, or tokio read.

Reply:
I now understand. Thank you for explaining it.

Reply:
See #17725